fenics install fails due to cmake python ply issue

I have come across a new error when installing fenics with spack. This seems like a similar issue to https://github.com/spack/spack/issues/13397.

I am installing spack with the following command:
`spack install fenics~hdf5 ^openmpi+pmi schedulers=slurm`

I run into cmake issues at the final step. Specifically it cannot locate py-ply and site.py
```
1 error found in build log:
     147      Ignoring EXACT since no version is requested.
     148    This warning is for project developers.  Use -Wno-dev to suppress it.
     149
     150    -- Found PythonLibs: /cm/shared/apps/spack-pkgs/linux-centos7-zen/gcc-8.2.0/python-3.8.6-u7px4ssttiqjhc6q2arrww4wqceck7nk/lib/libpython3.8.so (found versio
            n "3.8.6")
     151    ImportError: No module named site
     152    -- Could NOT find PY_ply (missing: PY_PLY_path)
  >> 153    CMake Error at CMakeLists.txt:438 (message):
     154      Required Python module 'ply' (http://www.dabeaz.com/ply/) could not be
     155      found.  Install ply or set DOLFIN_ENABLE_PYTHON to false.
     156
     157
     158    -- Configuring incomplete, errors occurred!
     159    See also "/home/apal/.spack_root/stage/spack-stage-fenics-2016.1.0-mgupwbojaxav4wuo63yfu6gv7oiy5jdn/spack-build-mgupwbo/CMakeFiles/CMakeOutput.log".

See build log for details:
  /home/apal/.spack_root/stage/spack-stage-fenics-2016.1.0-mgupwbojaxav4wuo63yfu6gv7oiy5jdn/spack-build-out.txt
```
I have verified that py-ply is in PYTHONPATH. I have explicitly spack-loaded python3 and py-ply.
```
[root@L01 tarballs]# spack find --loaded
==> 18 installed packages
-- linux-centos7-zen / gcc@8.2.0 --------------------------------
bzip2@1.0.8   gdbm@1.18.1   libbsd@0.10.0  libiconv@1.16  libxml2@2.9.10  openssl@1.1.1h  python@3.8.6  sqlite@3.33.0  xz@5.2.5
expat@2.2.10  gettext@0.21  libffi@3.3     libuuid@1.0.3  ncurses@6.2     py-ply@3.11     readline@8.0  tar@1.32       zlib@1.2.11
[root@L01 tarballs]#
[root@L01 tarballs]# python -c "import re, ply, site; print(re.compile('/__init__.py.*').sub('',ply.__file__))"
/cm/shared/apps/spack-pkgs/linux-centos7-zen/gcc-8.2.0/py-ply-3.11-fpyplf44xgseqlpml5vxvs4r7bxv4c53/lib/python3.8/site-packages/ply
[root@L01 tarballs]# python -c "import re, ply, site; print(re.compile('/__init__.py.*').sub('',site.__file__))"
/cm/shared/apps/spack-pkgs/linux-centos7-zen/gcc-8.2.0/python-3.8.6-u7px4ssttiqjhc6q2arrww4wqceck7nk/lib/python3.8/site.py
[root@L01 tarballs]# sed 's/:/\n/g' <<< "$PYTHONPATH"
/cm/shared/apps/spack-pkgs/linux-centos7-zen/gcc-8.2.0/py-ply-3.11-fpyplf44xgseqlpml5vxvs4r7bxv4c53/lib64/python3.8/site-packages
/cm/shared/apps/spack-pkgs/linux-centos7-zen/gcc-8.2.0/py-ply-3.11-fpyplf44xgseqlpml5vxvs4r7bxv4c53/lib/python3.8/site-packages
/cm/shared/apps/spack-pkgs/linux-centos7-zen/gcc-8.2.0/python-3.8.6-u7px4ssttiqjhc6q2arrww4wqceck7nk/lib/python3.8
```
Any ideas how to circumvent this issue? Thank you for your attention. Regards.

@scheibelp 
I was able to see the PYTHONPATH that spack uses during build-env
```
[root@L01 software]# spack build-env fenics | grep PYTHONPATH | sed 's/:/\n/g'
==> Warning: gcc@8.2.0 cannot build optimized binaries for "zen2". Using best target possible: "zen"
PYTHONPATH=/cm/shared/apps/spack-pkgs/fenics/2016.1.0-hhcr4l3qjmadirzbyzcksxlynzxqye55/lib/python3.8/site-packages
/cm/shared/apps/spack-pkgs/fenics/2016.1.0-hhcr4l3qjmadirzbyzcksxlynzxqye55/lib64/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-numpy/1.19.4-kqhki5eafkdrtt3xj3opla5dins7uqpg/lib/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-numpy/1.19.4-kqhki5eafkdrtt3xj3opla5dins7uqpg/lib64/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-cython/0.29.21-aj6nww3fm7udn7euq3uf3jtxaeq4ofey/lib/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-cython/0.29.21-aj6nww3fm7udn7euq3uf3jtxaeq4ofey/lib64/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-setuptools/50.3.2-lmbfsmftmbkpbhr6hxdo2herlq2xqbpn/lib/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-setuptools/50.3.2-lmbfsmftmbkpbhr6hxdo2herlq2xqbpn/lib64/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-ply/3.11-fpyplf44xgseqlpml5vxvs4r7bxv4c53/lib/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-ply/3.11-fpyplf44xgseqlpml5vxvs4r7bxv4c53/lib64/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-six/1.14.0-me3k65ic2t5it7npzrsaddtsxzgvmzxz/lib/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-six/1.14.0-me3k65ic2t5it7npzrsaddtsxzgvmzxz/lib64/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-sympy/1.4-kvteabovy66apdhesr7yrajlnqbmgfdu/lib/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-sympy/1.4-kvteabovy66apdhesr7yrajlnqbmgfdu/lib64/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-mpmath/1.1.0-srq4iwf4q7ki25t4lpjllf4j5chfuc2f/lib/python3.8/site-packages
/cm/shared/apps/spack-pkgs/py-mpmath/1.1.0-srq4iwf4q7ki25t4lpjllf4j5chfuc2f/lib64/python3.8/site-packages
```
It appears py-ply is being loaded. But cmake fails just as before:
```
1 error found in build log:
     147      Ignoring EXACT since no version is requested.
     148    This warning is for project developers.  Use -Wno-dev to suppress i
            t.
     149
     150    -- Found PythonLibs: /cm/shared/apps/spack-pkgs/python/3.8.6-u7px4s
            sttiqjhc6q2arrww4wqceck7nk/lib/libpython3.8.so (found version "3.8.
            6")
     151    ImportError: No module named site
     152    -- Could NOT find PY_ply (missing: PY_PLY_path)
  >> 153    CMake Error at CMakeLists.txt:438 (message):
     154      Required Python module 'ply' (http://www.dabeaz.com/ply/) could n
            ot be
     155      found.  Install ply or set DOLFIN_ENABLE_PYTHON to false.
     156
     157
     158    -- Configuring incomplete, errors occurred!
     159    See also "/home/apal/.spack_root/stage/spack-stage-fenics-2016.1.0-
            hhcr4l3qjmadirzbyzcksxlynzxqye55/spack-build-hhcr4l3/CMakeFiles/CMa
            keOutput.log".

See build log for details:
  /home/apal/.spack_root/stage/spack-stage-fenics-2016.1.0-hhcr4l3qjmadirzbyzcksxlynzxqye55/spack-build-out.txt

```

--- a/var/spack/repos/builtin/packages/fenics/package.py
+++ b/var/spack/repos/builtin/packages/fenics/package.py
@@ -122,7 +122,7 @@ class Fenics(CMakePackage):
     depends_on('py-sphinx@1.0.1:', when='+doc', type='build')
 
     def cmake_args(self):
-        return [
+        args = [
             self.define_from_variant('BUILD_SHARED_LIBS', 'shared'),
             self.define('DOLFIN_SKIP_BUILD_TESTS', True),
             self.define_from_variant('DOLFIN_ENABLE_OPENMP', 'openmp'),
@@ -145,6 +145,12 @@ class Fenics(CMakePackage):
             self.define_from_variant('DOLFIN_ENABLE_VTK', 'vtk'),
             self.define_from_variant('DOLFIN_ENABLE_ZLIB', 'zlib'),
         ]
+
+        if '+python' in self.spec:
+            args.append(self.define(
+                'PYTHON_EXECUTABLE', self.spec['python'].command.path))
+
+        return args
 
     # set environment for bulding python interface
     def setup_build_environment(self, env):

